internal/integration/bounded_harvester.go

```diff
@@ -0,0 +1,114 @@
+package integration
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/newrelic/newrelic-telemetry-sdk-go/telemetry"
+	"github.com/sirupsen/logrus"
+)
+
+const boundHarvesterDefaultPeriod = 3 * time.Second
+const boundHarvesterDefaultMetricCap = 1e5
+const boundHarvesterDefaultMinReportInterval = 100 * time.Millisecond
+
+// bindHarvester creates a boundedHarvester from an existing harvester.
+// It also returns a cancel channel to stop the periodic harvest goroutine.
+func bindHarvester(inner harvester, cfg BoundedHarvesterCfg) (harvester, chan struct{}) {
+	if _, ok := inner.(*telemetry.Harvester); ok {
+		logrus.Debug("using telemetry.Harvester as underlying harvester, make sure to set HarvestPeriod to 0")
+	}
+
+	if cfg.HarvestPeriod == 0 {
+		cfg.HarvestPeriod = boundHarvesterDefaultPeriod
+	}
+
+	if cfg.MetricCap == 0 {
+		cfg.MetricCap = boundHarvesterDefaultMetricCap
+	}
+
+	if cfg.MinReportInterval == 0 {
+		cfg.MinReportInterval = boundHarvesterDefaultMinReportInterval
+	}
+
+	h := &boundedHarvester{
+		BoundedHarvesterCfg: cfg,
+		mtx:                 sync.Mutex{},
+		inner:               inner,
+	}
+
+	cancel := make(chan struct{})
+	go h.periodicHarvest(cancel)
+	return h, cancel
+}
+
+// BoundedHarvesterCfg stores the configurable values for boundedHarvester
+type BoundedHarvesterCfg struct {
+	// Never report more often than MinReportInterval
+	MinReportInterval time.Duration
+
+	// Report when the number of stored metrics is greater than MetricCap
+	MetricCap int
+	// Also report at least once every HarvestPeriod
+	HarvestPeriod time.Duration
+}
+
+// boundedHarvester wraps another harvester and triggers its HarvestNow operation when a number of metrics have been
+// collected, or periodically every HarvestPeriod.
+// Additionally, it ensures that reports do not happen more often than MinReportInterval
+type boundedHarvester struct {
+	BoundedHarvesterCfg
+
+	mtx sync.Mutex
+
+	reportedMetrics int
+	lastReport      time.Time
+
+	inner harvester
+}
+
+// RecordMetric records the metric in the underlying harvester and reports all of them if needed
+func (h *boundedHarvester) RecordMetric(m telemetry.Metric) {
+	h.inner.RecordMetric(m)
+	h.reportIfNeeded(context.Background(), 1, false)
```

Comment:
does it make sense to check this for each metric?
each time we need to wait on the mux check and carry on, we could check each MinReportInterval instead

Reply:
I guess we would need to wait anyway to increment 	h.reportedMetrics += newMetrics

Reply:
I have also thought about this. Right now it is all done in the same function in order to keep the concurrency model simple, although it might not be the most performant option, specially if one of the calls ends up triggering a `HarvestNow`.

I'll give this a thought.

Reply:
This is now solved, `RecordMetric` no longer calls `reportIfNeeded`. Instead, the periodic routine is tasked with doing so.

This has its own pros and cons, perhaps being the most relevant one that the routine is now required for the boundedHarvester to function.